### PR TITLE
Feature adjoint transient optimization

### PR DESCRIPTION
**Pull request by sunethwarna**
This PR adds followings:

1. A response function which runs a method "Run", response value and its gradients need to be calculated inside this "Run" method and needs to be passed in a dictionary for the optimizer.
2. Restart capabilities for this response function is implemented.
3. External analyzer class to handle this response function is implemented in the test case (05 test case with an example)
4. The example combines AdjointFluidApplication and FluidDynamicsApplication to calculate shape sensitivities for a fluid problem. (This is for a transient example)

### CODE REVIEW

**Review by philbucher — COMMENTED** [review]
@sunethwarna you don't need the is_MPI flags from what I saw
you can get the barriers also through the ModelPart-Communicator
if you do sth in rank 0 this works both with and without MPI

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/impose_parabolic_inlet_process.py
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/impose_parabolic_inlet_process.py:1]: I think the InletProcess in Fluid is able to do what you do with this process
@jcotela @rubenzorrilla ?

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/impose_parabolic_inlet_process.py
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/impose_parabolic_inlet_process.py:16]: if you decide to keep this then please use the function form `kratos/utilities/variable_utils` for this

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py
Comment (msandre) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py:10]: Is it correct that `optimization_model_part` is not used?

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py
Comment (msandre) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py:48]: I think it is safe to call `AssembleCurrentData` without mpi (it does nothing). So you don't need the if statement.

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py
Comment (msandre) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py:60]: This does not work for MPI right? It would only return the response gradients from the first process.

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/analyzer_external_transient.py
Comment (msandre) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/analyzer_external_transient.py:25]: File name should be like class name: `external_transient_analyzer.py`

**Review by msandre — CHANGES_REQUESTED** [review]
I would wait for a review from @armingeiser or @dbaumgaertner on this, but here are my small comments.

File: applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/05_External_Fluid_Adjoint_Transient_Analyzer/base_case/response_analysis.py:46]: this if is not needed for example

**Comment by msandre** [comment]
Suneth, this must be quite outdated with current master. I think it's better to first discuss with structural optimization team before spending time to update/fix the branch. If we don't want it anymore then we should close it.

**Pull request closed without merging**
